src/coreclr/jit/assertionprop.cpp

```diff
@@ -4894,6 +4898,98 @@ bool Compiler::optNonNullAssertionProp_Ind(ASSERT_VALARG_TP assertions, GenTree*
     return false;
 }
 
+//------------------------------------------------------------------------
+// optAssertionIsNonHeap: see if we can prove a tree's value will be not GC-tracked
+//   based on assertions
+//
+// Arguments:
+//   op - tree to check
+//   assertions  - set of live assertions
+//   pVnBased - [out] set to true if value numbers were used
+//   pIndex - [out] the assertion used in the proof
+//
+// Returns:
+//   true if the tree's value will be not GC-tracked
+//
+// Notes:
+//   Sets "pVnBased" if the assertion is value number based. If no matching
+//    assertions are found from the table, then returns "NO_ASSERTION_INDEX."
+//
+//   If both VN and assertion table yield a matching assertion, "pVnBased"
+//   is only set and the return value is "NO_ASSERTION_INDEX."
+//
+bool Compiler::optAssertionIsNonHeap(GenTree*         op,
+                                     ASSERT_VALARG_TP assertions DEBUGARG(bool* pVnBased)
+                                         DEBUGARG(AssertionIndex* pIndex))
+{
+    bool vnBased = (!optLocalAssertionProp && vnStore->IsVNObjHandle(op->gtVNPair.GetConservative()));
+#ifdef DEBUG
+    *pIndex   = NO_ASSERTION_INDEX;
+    *pVnBased = vnBased;
+#endif
+
+    if (vnBased)
+    {
+        return true;
+    }
+
+    if (op->IsIconHandle(GTF_ICON_OBJ_HDL))
```

Comment:
nit: I'd move this check up
1) it's cheaper than `IsVNObjHandle`
2) It's not guaranteed for the icon handle to always have VN